In [663]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns

In [664]:
data_dict = pd.read_table("2025_Problem_C_Data\\data_dictionary.csv", encoding="Windows-1252",sep=",")

# 变量说明

In [665]:
data_dict.iloc[0:8,]

summerOly_medal_counts.csv                                     Unnamed: 1  \
0                  variables                                    explanation   
1                       Rank      Rank of country based on total medals won   
2                        NOC  Name of country as recorded for that Olympics   
3                       Gold       Number of Gold medals the country earned   
4                     Silver     Number of Silver medals the country earned   
5                     Bronze     Number of Bronze medals the country earned   
6                      Total      Number of Total medals the country earned   
7                       Year                                   Olympic year   

      Unnamed: 2  
0        example  
1           1, 2  
2  China, France  
3        0, 1, 2  
4        0, 1, 2  
5        0, 1, 2  
6        0, 1, 2  
7     1896, 1900

In [666]:
data_dict.iloc[11:14,]

summerOly_medal_counts.csv  \
11                  variables   
12                       Year   
13                       Host   

                                           Unnamed: 1  \
11                                        explanation   
12                                       Olympic year   
13  Location where olympics were held (city, country)   

                  Unnamed: 2  
11                   example  
12                1896, 1900  
13  St. Louis, United States

In [667]:
data_dict.iloc[17:36,]

summerOly_medal_counts.csv  \
17                                          variables   
18                                              Sport   
19                                         Discipline   
20                                               Code   
21                                               Year   
22  Data set of current and discontinued summer pr...   
23                                 Column Descriptors   
24                             Sport and Discipline:    
25                                             Code:    
26                             Sports Governing Body:   
27                                      1896 – 2028:    
28  Figures in each cell of the associated column ...   
29  * The 1906 Intercalated Olympic Games were to ...   
30  S1^ The World Baseball Softball Confederation,...   
31  S2^ AIBA/IBA was the recognised body of boxing...   
32  S3^ Sailing and rowing were included in the pr...   
33  S4^ At the time skateboarding was announced as...   
34  S5^ Before 1924, ice sports like figure skatin...   
35  Sourced from Wikipedia https://en.wikipedia.or...   

                                           Unnamed: 1  \
17                                        explanation   
18                                     Sport category   
19               Discipline within the sport category   
20                            Code for the discipline   
21        Number of events in disipline for that year   
22                                                NaN   
23                                                NaN   
24  The sports (and disciplines) make up the curre...   
25  Each discipline is marked with a unique 3-char...   
26  A listing of the sports governing bodies assoc...   
27  The four number code in each column represents...   
28                                                NaN   
29                                                NaN   
30                                                NaN   
31                                                NaN   
32                                                NaN   
33                                                NaN   
34                                                NaN   
35                                                NaN   

                   Unnamed: 2  
17                    example  
18          Aquatics, Cycling  
19  Artistic swimming, Diving  
20                   SWA, DIV  
21                       0, 1  
22                        NaN  
23                        NaN  
24                        NaN  
25                        NaN  
26                        NaN  
27                        NaN  
28                        NaN  
29                        NaN  
30                        NaN  
31                        NaN  
32                        NaN  
33                        NaN  
34                        NaN  
35                        NaN

In [668]:
data_dict.iloc[39:50,]

summerOly_medal_counts.csv                                Unnamed: 1  \
39                  variables                               explanation   
40                       Name                              Athlete name   
41                        Sex                            Sex of athlete   
42                       Team  Country name as recorded at the Olympics   
43                        NOC                    Name of Country (code)   
44                       Year                              Olympic year   
45                     Season                          Type of Olympics   
46                       City             City where Olympics were held   
47                      Sport                            Sport category   
48                      Event                    Event within the sport   
49                      Medal                              Medal earned   

                Unnamed: 2  
39                 example  
40     Cai Yun, Carl Lewis  
41                    F, M  
42           China, France  
43                CHN, FRA  
44              1896, 1900  
45                  Summer  
46          Beijing, Paris  
47       Aquatics, Cycling  
48  Gymnastics Men's Rings  
49          No medal, Gold

以下是对该文件内容的翻译：

---

### **文件 1：夏季奥运会奖牌统计 (summerOly_medal_counts.csv)**

| 变量名称 | 解释                               | 示例               |
|----------|------------------------------------|--------------------|
| Rank     | 国家奖牌总数排名                   | "1, 2"             |
| NOC      | 该奥运会记录的国家名称             | "中国, 法国"       |
| Gold     | 国家获得的金牌数量                 | "0, 1, 2"          |
| Silver   | 国家获得的银牌数量                 | "0, 1, 2"          |
| Bronze   | 国家获得的铜牌数量                 | "0, 1, 2"          |
| Total    | 国家获得的奖牌总数                 | "0, 1, 2"          |
| Year     | 奥运会年份                         | "1896, 1900"       |

---

### **文件 2：夏季奥运会举办地 (summerOly_hosts.csv)**

| 变量名称 | 解释                                       | 示例                      |
|----------|--------------------------------------------|---------------------------|
| Year     | 奥运会年份                                 | "1896, 1900"              |
| Host     | 奥运会举办地（城市，国家）                 | "圣路易斯, 美国"          |

---

### **文件 3：夏季奥运会项目 (summerOly_programs.csv)**

| 变量名称   | 解释                                                         | 示例                            |
|------------|--------------------------------------------------------------|---------------------------------|
| Sport      | 体育类别                                                     | "水上运动, 自行车"             |
| Discipline | 体育类别内的具体项目                                         | "艺术体操, 跳水"               |
| Code       | 项目的代码                                                   | "SWA, DIV"                     |
| Year       | 该项目在某年奥运会中的比赛数量                               | "0, 1"                         |

#### 数据说明：
- **体育和项目 (Sport and Discipline)**：包括当前和已取消的夏季奥运会官方项目，按国际奥委会（IOC）使用的名称按字母顺序排列。
- **代码 (Code)**：每个项目由 IOC 分配一个独特的 3 字符标识符代码。
- **体育管理机构**：列出与奥运会体育相关的国际管理机构。
- **年份 (1896-2028)**：列的四位数代表相关奥运会的年份。
- 表格中的数字表示该项目在该年举办的比赛数量；某些项目可能标记为展示项目（非正式项目）。

---

### **文件 4：夏季奥运会运动员数据 (summerOly_athletes.csv)**

| 变量名称 | 解释                                   | 示例                       |
|----------|----------------------------------------|----------------------------|
| Name     | 运动员姓名                             | "蔡赟, 卡尔·刘易斯"        |
| Sex      | 运动员性别                             | "女 (F), 男 (M)"           |
| Team     | 国家名称（奥运会记录的）               | "中国, 法国"               |
| NOC      | 国家代码                               | "CHN, FRA"                 |
| Year     | 奥运会年份                             | "1896, 1900"               |
| Season   | 奥运会类型                             | 夏季 (Summer)              |
| City     | 奥运会举办城市                         | "北京, 巴黎"               |
| Sport    | 体育类别                               | "水上运动, 自行车"         |
| Event    | 项目名称                               | "体操男子吊环"             |
| Medal    | 获得的奖牌类型                         | "无奖牌, 金牌 (Gold)"      |

---

### **备注**
1. **1906年奥运会**：原计划每四年举办一次，地位与正式奥运会相同，但仅在 1906 年于雅典举办了一次，之后取消。
2. **冬季项目迁移**：在 1924 年之前，一些冰上项目（如花样滑冰和冰球）在夏季奥运会中举办，1924 年起转移至冬奥会。
3. **管理机构变化**：如拳击的管理机构从 2023 年起由 AIBA 改为 IBA。

---

# 获取分析历史奖牌数据

In [669]:
history_medal_counts = pd.read_csv("2025_Problem_C_Data\\summerOly_medal_counts.csv", encoding="Windows-1252",sep=",")

In [670]:
history_medal_counts

Rank                   NOC  Gold  Silver  Bronze  Total  Year
0        1         United States    11       7       2     20  1896
1        2                Greece    10      18      19     47  1896
2        3               Germany     6       5       2     13  1896
3        4                France     5       4       2     11  1896
4        5         Great Britain     2       3       2      7  1896
...    ...                   ...   ...     ...     ...    ...   ...
1430    84                 Qatar     0       0       1      1  2024
1431    84  Refugee Olympic Team     0       0       1      1  2024
1432    84             Singapore     0       0       1      1  2024
1433    84              Slovakia     0       0       1      1  2024
1434    84                Zambia     0       0       1      1  2024

[1435 rows x 7 columns]

# 对于历史上已经消失或者合并的国家的数据进行清洗

In [671]:
# 获取国家的名称
noc_names = history_medal_counts["NOC"].unique()
print(noc_names)

['United States' 'Greece' 'Germany' 'France' 'Great Britain' 'Hungary'
 'Austria' 'Australia' 'Denmark' 'Switzerland' 'Mixed team' 'Belgium'
 'Italy' 'Cuba' 'Canada' 'Spain' 'Luxembourg' 'Norway' 'Netherlands'
 'India' 'Bohemia' 'Sweden' 'Australasia' 'Russian Empire' 'Finland'
 'South Africa' 'Estonia' 'Brazil' 'Japan' 'Czechoslovakia' 'New Zealand'
 'Yugoslavia' 'Argentina' 'Uruguay' 'Poland' 'Haiti' 'Portugal' 'Romania'
 'Egypt' 'Ireland' 'Chile' 'Philippines' 'Mexico' 'Latvia' 'Turkey'
 'Jamaica' 'Peru' 'Ceylon' 'Trinidad and Tobago' 'Panama' 'South Korea'
 'Iran' 'Puerto Rico' 'Soviet Union' 'Lebanon' 'Bulgaria' 'Venezuela'
 'United Team of Germany' 'Iceland' 'Pakistan' 'Bahamas' 'Ethiopia'
 'Formosa' 'Ghana' 'Morocco' 'Singapore' 'British West Indies' 'Iraq'
 'Tunisia' 'Kenya' 'Nigeria' 'East Germany' 'West Germany' 'Mongolia'
 'Uganda' 'Cameroon' 'Taiwan' 'North Korea' 'Colombia' 'Niger' 'Bermuda'
 'Thailand' 'Zimbabwe' 'Tanzania' 'Guyana' 'China' 'Ivory Coast' 'Syria'
 'Algeria

In [672]:
# Step 1: Redefine the removal and merge rules based on updated requirements
# Countries/regions to remove entirely
final_remove_nocs = [
    "Bohemia", "Australasia", "Russian Empire", "Czechoslovakia",
    "Unified Team", "Mixed team", "East Germany", "West Germany",
    "United Team of Germany", "FR Yugoslavia", "Serbia and Montenegro",
    "Soviet Union"
]

# Update merge mappings
final_merge_nocs = {
    "Formosa": "Chinese Taipei"  # Merge Formosa with Chinese Taipei
    # Note: China, Hong Kong, and Macau remain separate
}

# Step 2: Start cleaning process from original data
final_cleaned_data = history_medal_counts[~history_medal_counts["NOC"].isin(final_remove_nocs)]

# Replace NOC names based on updated merge mapping
final_cleaned_data["NOC"] = final_cleaned_data["NOC"].replace(final_merge_nocs)

# Step 3: Recalculate totals and ranks with the updated cleaning rules
final_cleaned_data = (
    final_cleaned_data.groupby(["Year", "NOC"], as_index=False)
    .agg({"Gold": "sum", "Silver": "sum", "Bronze": "sum"})
)
final_cleaned_data["Total"] = final_cleaned_data["Gold"] + final_cleaned_data["Silver"] + final_cleaned_data["Bronze"]

# Rank NOCs within each year
final_cleaned_data["Rank"] = (
    final_cleaned_data.sort_values(["Year", "Total", "Gold", "Silver"], ascending=[True, False, False, False])
    .groupby("Year")
    .cumcount() + 1
)

# 保存数据
final_cleaned_data.head()
cleaned_medal_data = final_cleaned_data
cleaned_medal_data.to_csv("2025_Problem_C_Data_cleaned\\cleaned_medal_data.csv", index=False)


C:\Users\Tieti_Tage\AppData\Local\Temp\ipykernel_29004\3302144721.py:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_cleaned_data["NOC"] = final_cleaned_data["NOC"].replace(final_merge_nocs)


In [673]:
# # 遍历分组
# for name, group in final_cleaned_data.groupby("NOC"):
#     # 绘制每个 NOC 的奖牌柱状图
#     group.plot(x="Year", y=["Gold", "Silver", "Bronze"], kind="bar", title=name)
#     plt.tight_layout()  # 调整布局
#     plt.show()  # 显示每个图


- 清洗后得到的各个国家和地区的历史奖牌的柱状图如图

# 读取项目表

In [674]:
program_data = pd.read_csv("2025_Problem_C_Data\\summerOly_programs.csv", encoding="latin1",sep=",")

In [675]:
program_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 74 entries, 0 to 73
Data columns (total 35 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Sport                  74 non-null     object 
 1   Discipline             72 non-null     object 
 2   Code                   74 non-null     object 
 3   Sports Governing Body  74 non-null     object 
 4   1896                   74 non-null     object 
 5   1900                   74 non-null     object 
 6   1904                   74 non-null     object 
 7   1906*                  74 non-null     int64  
 8   1908                   74 non-null     int64  
 9   1912                   74 non-null     object 
 10  1920                   74 non-null     int64  
 11  1924                   74 non-null     object 
 12  1928                   72 non-null     object 
 13  1932                   72 non-null     object 
 14  1936                   72 non-null     object 
 15  1948    

In [676]:
print(pd.unique(program_data["Discipline"]))

['Artistic Swimming' 'Diving' 'Marathon Swimming' 'Swimming' 'Water Polo'
 'Archery' 'Athletics' 'Badminton' 'Baseball' 'Softball' '3x3'
 'Basketball' 'Basque Pelota' 'Boxing' 'Breaking' 'Sprint' 'Slalom'
 'Cricket' 'Croquet' 'BMX Freestyle' 'BMX Racing' 'Mountain Bike' 'Road'
 'Track' 'Dressage' 'Eventing' 'Jumping' 'Vaulting' 'Driving' 'Fencing'
 'Field hockey' 'Flag football' 'Football' 'Golf' 'Artistic' 'Rhythmic'
 'Trampoline' 'Indoor' 'Field' 'Jeu de Paume' 'Judo' 'Karate' 'Sixes' nan
 'Polo' 'Rackets' 'Roque' 'Coastal' 'Rowing' 'Sevens' 'Union' 'Sailing'
 'Shooting' 'Skateboarding' 'Sport Climbing' 'Squash' 'Surfing'
 'Table Tennis' 'Taekwondo' 'Tennis' 'Triathlon' 'Tug of War' 'Beach'
 'Weightlifting' 'Freestyle' 'Greco-Roman' 'Figure' 'Ice Hockey'
 'Total events' 'Total disciplines' 'Total sports']


In [677]:
# 清除不太可能存在于2028奥运会当中的项目
unlikely_events = [
    "Boxing", "Weightlifting", "Breaking",
    "Croquet", "Softball", "Lacrosse", 
    "Squash", "Cricket","Water Motorsports","Skating","Ice Hockey"
]

# Remove these unlikely events from the dataset
filtered_program_data = program_data[~program_data['Sport'].isin(unlikely_events)]
filtered_program_data.info()



<class 'pandas.core.frame.DataFrame'>
Index: 63 entries, 0 to 73
Data columns (total 35 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Sport                  63 non-null     object 
 1   Discipline             62 non-null     object 
 2   Code                   63 non-null     object 
 3   Sports Governing Body  63 non-null     object 
 4   1896                   63 non-null     object 
 5   1900                   63 non-null     object 
 6   1904                   63 non-null     object 
 7   1906*                  63 non-null     int64  
 8   1908                   63 non-null     int64  
 9   1912                   63 non-null     object 
 10  1920                   63 non-null     int64  
 11  1924                   63 non-null     object 
 12  1928                   63 non-null     object 
 13  1932                   63 non-null     object 
 14  1936                   63 non-null     object 
 15  1948         

1. **拳击（Boxing）**：由于国际拳击联合会的管理问题，拳击项目未被列入2028年洛杉矶奥运会的初始项目名单。尽管如此，如果相关问题得到解决，拳击仍有可能被重新纳入。 ([澎湃新闻](https://www.thepaper.cn/newsDetail_forward_15781574?utm_source=chatgpt.com))

2. **举重（Weightlifting）**：因反兴奋剂方面的历史问题，举重项目也未被列入初始名单。国际举重联合会需要进行改革，以确保该项目能够回归奥运会。 ([澎湃新闻](https://www.thepaper.cn/newsDetail_forward_15781574?utm_source=chatgpt.com))


4. **霹雳舞（Breaking）**：尽管在2024年巴黎奥运会上首次亮相，但霹雳舞未被纳入2028年洛杉矶奥运会的项目。 ([Diario AS](https://as.com/juegos_olimpicos/proxima-parada-hollywood-7000-millones-sin-coches-criquet-n/?utm_source=chatgpt.com))

5. **槌球（Croquet）**、**壘球（Softball）**、**长曲棍球（Lacrosse）**、**壁球（Squash）**、**板球（Cricket）**：这些项目在历史上曾出现在奥运会上，但在近期的奥运会中并未包含。然而，洛杉矶奥组委提议将棒垒球、板球、腰旗橄榄球、棍网球和壁球作为新增大项纳入2028年奥运会。 


In [678]:
# 填充缺失值
# 仅有Modern Pentathlon项目不存在归属，直接使用其名字作为discipline
filtered_program_data.fillna("Modern Pentathlon", inplace=True)

C:\Users\Tieti_Tage\AppData\Local\Temp\ipykernel_29004\1671211733.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_program_data.fillna("Modern Pentathlon", inplace=True)


In [679]:
# 转换类型
# 遍历所有列，将日期列的数据类型转换为整数
for col in filtered_program_data.columns:
    if col not in ["Discipline", "Sport", "Sports Governing Body","Code"]:
        # 将列数据转换为整数，跳过可能导致错误的数据
        filtered_program_data[col] = pd.to_numeric(filtered_program_data[col], errors='coerce').fillna(0).astype(int)

C:\Users\Tieti_Tage\AppData\Local\Temp\ipykernel_29004\3356668179.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_program_data[col] = pd.to_numeric(filtered_program_data[col], errors='coerce').fillna(0).astype(int)
C:\Users\Tieti_Tage\AppData\Local\Temp\ipykernel_29004\3356668179.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_program_data[col] = pd.to_numeric(filtered_program_data[col], errors='coerce').fillna(0).astype(int)
C:\Users\Tieti_Tage\AppData\Local\Temp\ipykernel_29004\3

In [680]:
filtered_program_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 63 entries, 0 to 73
Data columns (total 35 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   Sport                  63 non-null     object
 1   Discipline             63 non-null     object
 2   Code                   63 non-null     object
 3   Sports Governing Body  63 non-null     object
 4   1896                   63 non-null     int32 
 5   1900                   63 non-null     int32 
 6   1904                   63 non-null     int32 
 7   1906*                  63 non-null     int32 
 8   1908                   63 non-null     int32 
 9   1912                   63 non-null     int32 
 10  1920                   63 non-null     int32 
 11  1924                   63 non-null     int32 
 12  1928                   63 non-null     int32 
 13  1932                   63 non-null     int32 
 14  1936                   63 non-null     int32 
 15  1948                   63 non-

In [681]:
filtered_program_data.head(10)

Sport         Discipline Code Sports Governing Body  1896  \
0               Aquatics  Artistic Swimming  SWA        World Aquatics     0   
1               Aquatics             Diving  DIV        World Aquatics     0   
2               Aquatics  Marathon Swimming  OWS        World Aquatics     0   
3               Aquatics           Swimming  SWM        World Aquatics     4   
4               Aquatics         Water Polo  WPO        World Aquatics     0   
5                Archery            Archery  ARC         World Archery     0   
6              Athletics          Athletics  ATH       World Athletics    12   
7              Badminton          Badminton  BDM                   BWF     0   
8  Baseball and Softball           Baseball  BSB              WBSC[s1]     0   
9  Baseball and Softball           Softball  SBL              WBSC[s1]     0   

   1900  1904  1906*  1908  1912  ...  1988  1992  1996  2000  2004  2008  \
0     0     0      0     0     0  ...     2     2     1     2     2     2   
1     0     2      1     2     4  ...     4     4     4     8     8     8   
2     0     0      0     0     0  ...     0     0     0     0     0     2   
3     7     9      4     6     9  ...    31    31    32    32    32    32   
4     1     1      0     1     1  ...     1     1     1     2     2     2   
5     7     6      0     3     0  ...     4     4     4     4     4     4   
6    23    25     21    26    30  ...    42    43    44    46    46    47   
7     0     0      0     0     0  ...     0     4     5     5     5     5   
8     0     0      0     0     0  ...     0     1     1     1     1     1   
9     0     0      0     0     0  ...     0     0     1     1     1     1   

   2012  2016  2020  2024  
0     2     2     2     2  
1     8     8     8     8  
2     2     2     2     2  
3    32    32    35    35  
4     2     2     2     2  
5     4     4     5     5  
6    47    47    48    48  
7     5     5     5     5  
8     0     0     1     0  
9     0     0     1     0  

[10 rows x 35 columns]

In [682]:
# 保存清洗后的数据
filtered_program_data.to_csv("2025_Problem_C_Data_cleaned\\filtered_program_data.csv", encoding="utf-8",index=False)


In [683]:
# 只保留1990年以来的奥运会的项目情况
# 删除第 4 到第 25 列
filtered_program_data.drop(columns=filtered_program_data.columns[4:26], inplace=True)
filtered_program_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 63 entries, 0 to 73
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   Sport                  63 non-null     object
 1   Discipline             63 non-null     object
 2   Code                   63 non-null     object
 3   Sports Governing Body  63 non-null     object
 4   1992                   63 non-null     int32 
 5   1996                   63 non-null     int32 
 6   2000                   63 non-null     int32 
 7   2004                   63 non-null     int32 
 8   2008                   63 non-null     int32 
 9   2012                   63 non-null     int32 
 10  2016                   63 non-null     int32 
 11  2020                   63 non-null     int32 
 12  2024                   63 non-null     int32 
dtypes: int32(9), object(4)
memory usage: 4.7+ KB


C:\Users\Tieti_Tage\AppData\Local\Temp\ipykernel_29004\2660240754.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_program_data.drop(columns=filtered_program_data.columns[4:26], inplace=True)


In [684]:
rows_with_na = filtered_program_data[filtered_program_data.isna().any(axis=1)]
print("含缺失值的行：")
print(rows_with_na)

含缺失值的行：
Empty DataFrame
Columns: [Sport, Discipline, Code, Sports Governing Body, 1992, 1996, 2000, 2004, 2008, 2012, 2016, 2020, 2024]
Index: []


# 读取东道主的数据

In [685]:
host_data = pd.read_csv("2025_Problem_C_Data\\summerOly_hosts.csv")

In [686]:
host_data

Year                                               Host
0   1896                                     Athens, Greece
1   1900                                      Paris, France
2   1904                           St. Louis, United States
3   1908                             London, United Kingdom
4   1912                                  Stockholm, Sweden
5   1916          Cancelled (WWI – Berlin had been awarded)
6   1920                                   Antwerp, Belgium
7   1924                                      Paris, France
8   1928                             Amsterdam, Netherlands
9   1932                         Los Angeles, United States
10  1936                                    Berlin, Germany
11  1940          Cancelled (WWII – Tokyo had been awarded)
12  1944         Cancelled (WWII – London had been awarded)
13  1948                             London, United Kingdom
14  1952                                  Helsinki, Finland
15  1956                               Melbourne, Australia
16  1960                                        Rome, Italy
17  1964                                       Tokyo, Japan
18  1968                                Mexico City, Mexico
19  1972                               Munich, West Germany
20  1976                                   Montreal, Canada
21  1980                               Moscow, Soviet Union
22  1984                         Los Angeles, United States
23  1988                                 Seoul, South Korea
24  1992                                   Barcelona, Spain
25  1996                             Atlanta, United States
26  2000                                  Sydney, Australia
27  2004                                     Athens, Greece
28  2008                                     Beijing, China
29  2012                             London, United Kingdom
30  2016                             Rio de Janeiro, Brazil
31  2020   Tokyo, Japan (postponed to 2021 due to the co...
32  2024                                      Paris, France
33  2028                         Los Angeles, United States
34  2032                                Brisbane, Australia

# 读取并清洗运动员的数据

In [687]:
athletics_data = pd.read_csv("2025_Problem_C_Data\\summerOly_athletes.csv")

In [688]:
athletics_data

Name Sex               Team  NOC  Year         City  \
0                   A Dijiang   M              China  CHN  1992    Barcelona   
1                    A Lamusi   M              China  CHN  2012       London   
2                 Gunnar Aaby   M            Denmark  DEN  1920    Antwerpen   
3                 Edgar Aabye   M     Denmark/Sweden  DEN  1900        Paris   
4       Cornelia (-strannood)   F        Netherlands  NED  1932  Los Angeles   
...                       ...  ..                ...  ...   ...          ...   
252560             Sefora Ada   F  Equatorial Guinea  GEQ  2024        Paris   
252561        Emanuela Liuzzi   F              Italy  ITA  2024        Paris   
252562           Isayah Boers   M        Netherlands  NED  2024        Paris   
252563            Kevin Staut   M             France  FRA  2024        Paris   
252564        Charlie Carvell   M      Great Britain  GBR  2024        Paris   

             Sport                         Event     Medal  
0       Basketball   Basketball Men's Basketball  No medal  
1             Judo  Judo Men's Extra-Lightweight  No medal  
2         Football       Football Men's Football  No medal  
3       Tug-Of-War   Tug-Of-War Men's Tug-Of-War      Gold  
4        Athletics  Athletics Women's 100 metres  No medal  
...            ...                           ...       ...  
252560   Athletics                  Women's 100m  No medal  
252561   Wrestling        Women's Freestyle 50kg  No medal  
252562   Athletics          4 x 400m Relay Mixed      Gold  
252563  Equestrian                  Jumping Team    Bronze  
252564   Athletics          Men's 4 x 400m Relay    Bronze  

[252565 rows x 9 columns]

In [689]:
athletics_data.isna().sum()

Name     0
Sex      0
Team     0
NOC      0
Year     0
City     0
Sport    0
Event    0
Medal    0
dtype: int64

In [690]:
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import matplotlib.pyplot as plt

# Step 1: 加载数据
athletes = pd.read_csv('2025_Problem_C_Data/summerOly_athletes.csv', encoding='latin1')
hosts = pd.read_csv('2025_Problem_C_Data/summerOly_hosts.csv', encoding='latin1')


# Step 2: 清洗数据
# 移除诡异的字符
hosts['Host'] = hosts['Host'].str.replace(r'[Â\xa0]', '', regex=True).str.strip()

In [691]:

# 移除无奖牌的数据
athletes = athletes[athletes['Medal'] != 'No medal']

# Group data by Team, Year, Sport, and Medal, then count the occurrences
grouped_data = athletes.groupby(['Team', 'Year', 'Sport', 'Medal']).size().reset_index(name='Count')

# Pivot the data to get Medal types as columns
pivot_table = grouped_data.pivot_table(
    index=['Team', 'Year', 'Sport'],
    columns='Medal',
    values='Count',
    fill_value=0
).reset_index()


In [692]:
pivot_table.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7018 entries, 0 to 7017
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Team    7018 non-null   object 
 1   Year    7018 non-null   int64  
 2   Sport   7018 non-null   object 
 3   Bronze  7018 non-null   float64
 4   Gold    7018 non-null   float64
 5   Silver  7018 non-null   float64
dtypes: float64(3), int64(1), object(2)
memory usage: 329.1+ KB


In [693]:
print(pivot_table.head(4))

Medal                   Team  Year                  Sport  Bronze  Gold  \
0      A North American Team  1900                   Polo     4.0   0.0   
1                        AIN  2024                 Rowing     0.0   0.0   
2                        AIN  2024                 Tennis     0.0   0.0   
3                        AIN  2024  Trampoline Gymnastics     0.0   1.0   

Medal  Silver  
0         0.0  
1         1.0  
2         2.0  
3         1.0  


In [694]:
# 删除1990前的数据
pivot_table = pivot_table[pivot_table["Year"]>1990]

In [695]:
pivot_table.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3672 entries, 1 to 7016
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Team    3672 non-null   object 
 1   Year    3672 non-null   int64  
 2   Sport   3672 non-null   object 
 3   Bronze  3672 non-null   float64
 4   Gold    3672 non-null   float64
 5   Silver  3672 non-null   float64
dtypes: float64(3), int64(1), object(2)
memory usage: 200.8+ KB


In [696]:
# 标准化国家名称
pivot_table['Team'] = pivot_table['Team'].str.replace(r'-\d+$', '', regex=True)


In [697]:
# 保存表格
output_path = '2025_Problem_C_Data_cleaned/olympic_medal_analysis.csv'
pivot_table.to_csv(output_path, index=False)

# 合并表格数据,将cleaned_medal_data当中的数据与运动项目的数据进行合并

In [707]:
import pandas as pd

# 加载数据文件
file_path_original = '2025_Problem_C_Data_cleaned/olympic_medal_analysis.csv'
file_path_new = '2025_Problem_C_Data_cleaned/cleaned_medal_data.csv'
data = pd.read_csv(file_path_original)
new_data = pd.read_csv(file_path_new)

# 步骤 1: 数据清洗
# 确保存在 'Total' 列，如果没有则通过求和 'Gold', 'Silver', 'Bronze' 列创建
if 'Total' not in data.columns:
    data['Total'] = data[['Gold', 'Silver', 'Bronze']].sum(axis=1)

# 标准化国家名称，移除例如 '-1' 的后缀
data['Team'] = data['Team'].str.replace(r'-\d+$', '', regex=True)
data.rename(columns={'Team': 'NOC'}, inplace=True)

# 步骤 2: 创建按国家、年份和运动项目统计的奖牌数据透视表
# 创建金牌数据表
gold_pivot = data.pivot_table(
    index=['NOC', 'Year'], 
    columns='Sport', 
    values='Gold', 
    aggfunc='sum', 
    fill_value=0
)
gold_pivot = gold_pivot.drop(columns=unlikely_events, errors='ignore')  # 移除不可能的项目
gold_summary = data.groupby(['NOC', 'Year'])[['Gold']].sum()
gold_data = pd.concat([gold_summary, gold_pivot], axis=1).reset_index()
gold_file_path = '2025_Problem_C_Data_cleaned/gold_medal_data.csv'
gold_data.to_csv(gold_file_path, index=False)

# 创建银牌数据表
silver_pivot = data.pivot_table(
    index=['NOC', 'Year'], 
    columns='Sport', 
    values='Silver', 
    aggfunc='sum', 
    fill_value=0
)
silver_pivot = silver_pivot.drop(columns=unlikely_events, errors='ignore')
silver_summary = data.groupby(['NOC', 'Year'])[['Silver']].sum()
silver_data = pd.concat([silver_summary, silver_pivot], axis=1).reset_index()
silver_file_path = '2025_Problem_C_Data_cleaned/silver_medal_data.csv'
silver_data.to_csv(silver_file_path, index=False)

# 创建铜牌数据表
bronze_pivot = data.pivot_table(
    index=['NOC', 'Year'], 
    columns='Sport', 
    values='Bronze', 
    aggfunc='sum', 
    fill_value=0
)

bronze_pivot = bronze_pivot.drop(columns=unlikely_events, errors='ignore')
bronze_summary = data.groupby(['NOC', 'Year'])[['Bronze']].sum()
bronze_data = pd.concat([bronze_summary, bronze_pivot], axis=1).reset_index()
bronze_file_path = '2025_Problem_C_Data_cleaned/bronze_medal_data.csv'
bronze_data.to_csv(bronze_file_path, index=False)

# 输出保存的文件路径
print("Gold medals saved to:", gold_file_path)
print("Silver medals saved to:", silver_file_path)
print("Bronze medals saved to:", bronze_file_path)

Gold medals saved to: 2025_Problem_C_Data_cleaned/gold_medal_data.csv
Silver medals saved to: 2025_Problem_C_Data_cleaned/silver_medal_data.csv
Bronze medals saved to: 2025_Problem_C_Data_cleaned/bronze_medal_data.csv
